mapie/calibrators/base.py

```diff
@@ -0,0 +1,85 @@
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from typing import List, Optional
+
+from sklearn.base import BaseEstimator
+
+from mapie._typing import ArrayLike, NDArray
+
+
+class BaseCalibrator(BaseEstimator, metaclass=ABCMeta):
+    """
+    Base abstract class for the calibrators.
+
+    The ``BaseCalibrator`` subclasses should have at least two methods:
+
+    - ``fit`` : Fit the calibrator to estimate the conformity scores
+        quantiles.
+
+    -  ``predict`` : Predict the conformity score quantiles.
+
+    Attributes
+    ----------
+    fit_attributes: Optional[List[str]]
+        Name of attributes set during the ``fit`` method, and required to call
+        ``predict``.
+    """
+
+    fit_attributes: List[str]
+    sym: bool
+    alpha: Optional[float]
+    random_state: Optional[int]
+
+    @abstractmethod
+    def fit(
+        self,
+        X_calib: ArrayLike,
```

Comment:
Not sure about X_calib, maybe keep it general and just say `X`

Reply:
I wanted (and was encouraged) to allow the users to build any custom calibrators they want. So the calibration can take any of the following arguments:
```python
[
    X, y, sample_weight, groups,
    y_pred_calib, conformity_scores_calib,
    X_train, y_train, z_train, sample_weight_train, train_index,
    X_calib, y_calib, z_calib, sample_weight_calib, calib_index,
] + *kwargs + any of the `self` attributes
```
For instance, the ``CQR`` method would require ``X_train, y_train, X_calib, y_calib`` and even ``predictor``(from the self) if we don't specify the quantile regressor of the calibrator, but just use the same as the predictor (as it is done right know in ``MapieQuantileRegresor``).

Knowing that, I can't just rename ``X_calib`` to ``X``.  
But maybe we could still allow generalisation to other calibrator in a better way ?